config/routes.rb

```diff
@@ -1,4 +1,19 @@
 Rails.application.routes.draw do
+
+  root 'welcome#index'
+  post '/books/:id/upvote' => 'books#upvote', as: :upvote_books
```

Comment:
Since you use `resources :books` down below, you don't need to write this out like this. You could instead write:

`resources :books do
  patch 'upvote'
end`

This will automatically generate the route and route helper like you have it written here.